staging/src/k8s.io/apiserver/pkg/apis/config/validation/validation.go

```diff
@@ -0,0 +1,208 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// Package validation validates EncryptionConfiguration.
+package validation
+
+import (
+	"encoding/base64"
+	"fmt"
+	"net/url"
+
+	"k8s.io/apimachinery/pkg/util/validation/field"
+	"k8s.io/apiserver/pkg/apis/config"
+)
+
+const (
+	moreThanOneElementErr = "more than one provider specified in a single element, should split into different list elements"
+)
+
+var (
+	aesKeySizes              = []int{16, 24, 32}
+	secretBoxKeySizes        = []int{32}
+	root                     = field.NewPath("EncryptionConfiguration")
+	resource                 = root.Child("Resource")
+	keyLenErrFmt             = "secret is not of the expected length, got %d, expected one of %v"
+	unsupportedSchemeErrFmt  = "unsupported scheme %q for KMS provider, only unix is supported"
+	atLeastOneRequiredErrFmt = "at least one %s is required"
+	mandatoryFieldErrFmt     = "%s is a mandatory field for a %s"
+	base64EncodingErr        = "Secrets must be base64 encoded"
+	zeroOrNegativeErrFmt     = "%s should be a positive value"
```

Comment:
optional: it would be cleaner to inspect only `ErrorType` and `Field` in https://godoc.org/k8s.io/apimachinery/pkg/util/validation/field#Error

Reply:
I agree it would be cleaner, but I am worried how tests would differentiate produced errors.
Most ErrorTypes are the same - Invalid.